packages/react-transport-dom-webpack/src/ReactFlightWebpackPlugin.js

```diff
@@ -8,17 +8,170 @@
+  }
+}
+
+// This is the module that will be used to anchor all client references to.
+// I.e. it will have all the client files as async deps from this point on.
+// We use the Flight client implementation because you can't get to these
+// without the client runtime so it's the first time in the loading sequence
+// you might want them.
+const clientFileName = require.resolve('../');
+
+type ClientReferenceSearchPath = {
+  directory: string,
+  recursive?: boolean,
+  include: RegExp,
+  exclude?: RegExp,
+};
+
+type ClientReferencePath = string | ClientReferenceSearchPath;
+
+type Options = {
+  isServer: boolean,
+  clientReferences?: ClientReferencePath | $ReadOnlyArray<ClientReferencePath>,
+  chunkName?: string,
+};
+
+const PLUGIN_NAME = 'React Transport Plugin';
+
 export default class ReactFlightWebpackPlugin {
-  constructor(options: {isServer: boolean}) {}
+  clientReferences: $ReadOnlyArray<ClientReferencePath>;
+  chunkName: string;
+  constructor(options: Options) {
+    if (!options || typeof options.isServer !== 'boolean') {
+      throw new Error(
+        PLUGIN_NAME + ': You must specify the isServer option as a boolean.',
+      );
+    }
+    if (options.isServer) {
+      throw new Error('TODO: Implement the server compiler.');
+    }
+    if (!options.clientReferences) {
+      this.clientReferences = [
+        {
+          directory: '.',
+          recursive: true,
+          include: /\.client\.(js|ts|jsx|tsx)$/,
+        },
+      ];
+    } else if (
+      typeof options.clientReferences === 'string' ||
+      !Array.isArray(options.clientReferences)
+    ) {
+      this.clientReferences = [(options.clientReferences: $FlowFixMe)];
+    } else {
+      this.clientReferences = options.clientReferences;
+    }
+    if (typeof options.chunkName === 'string') {
+      this.chunkName = options.chunkName;
+      if (!/\[(index|request)\]/.test(this.chunkName)) {
+        this.chunkName += '[index]';
+      }
+    } else {
+      this.chunkName = 'client[index]';
+    }
+  }
 
   apply(compiler: any) {
-    compiler.hooks.emit.tap('React Transport Plugin', compilation => {
+    const run = (params, callback) => {
+      // First we need to find all client files on the file system. We do this early so
+      // that we have them synchronously available later when we need them. This might
+      // not be needed anymore since we no longer need to compile the module itself in
+      // a special way. So it's probably better to do this lazily and in parallel with
+      // other compilation.
+      const contextResolver = compiler.resolverFactory.get('context', {});
+      this.resolveAllClientFiles(
+        compiler.context,
+        contextResolver,
+        compiler.inputFileSystem,
+        compiler.createContextModuleFactory(),
+        (err, resolvedClientReferences) => {
+          if (err) {
+            callback(err);
+            return;
+          }
+          compiler.hooks.compilation.tap(
```

Comment:
Is it safe to `tap` asynchronously like this in a callback? Is there a chance that we might miss events while `resolveAllClientFiles` is still collecting them?

Reply:
I believe this is a blocking phase so Webpack doesn’t proceed until we’ve invoked the callback.

That’s also why this approach is suboptimal since it doesn’t allow for other parallelism in the meantime.

That was a limitation in the previous approach that needed to change/split the files.

We could put this elsewhere but it’s not broken.